# Tarea 1: Introducción, Vector Space Models, Information Retrieval y Language Models</h1>
**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2024)**

## Tarjeta de identificación

**Nombres:** ```Diego Acevedo, Benjamín Aguilar y Luis Montero```

**Fecha límite de entrega 📆:** 10/04.

**Tiempo estimado de dedicación:** 4 horas


## Instrucciones
Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en **Information Retrieval (IR)**, **Vector Space Models** y **Language Models**. Si aún no has visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP.

* La tarea es en **grupo** (maximo hasta 3 personas).
* La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
* El formato de entrega es este mismo Jupyter Notebook.
* Al momento de la revisión su código será ejecutado. Por favor verifiquen que su entrega no tenga errores de compilación.
* Completar la tarjeta de identificación. Sin ella no podrá tener nota.

## Material de referencia

Diapositivas del curso 📄
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    
- [Probabilistic Language Models](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-PLM.pdf)

Videos del curso 📺
- Introducción  [Parte 1](https://www.youtube.com/watch?v=HEKTNOttGvU)  [Parte 2](https://www.youtube.com/watch?v=P8cwnI-f-Kg)

- Information Retrieval [Parte 1](https://www.youtube.com/watch?v=FXIVClF370w&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) [Parte 2](https://www.youtube.com/watch?v=f8nG1EMmPZk&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3)
- Probabilistic Language Models [Parte 1](https://www.youtube.com/watch?v=9E2jJ6kcb4Y&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) [Parte 2](https://www.youtube.com/watch?v=ZWqbEQXLra0&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=5) [Parte 3](https://www.youtube.com/watch?v=tsumFqwFlaA&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=6) [Parte 4](https://www.youtube.com/watch?v=s3TWdv4sqkg&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJq)

## P1. Tokenización

En el primer ejercicio veremos la dificultad de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo.

In [1]:
# En caso de desarrollar la tarea desde colab, con el siguiente código podemos cargar los archivos desde drive:

try:
    from google.colab import drive

    drive.mount("/content/drive", force_remount=True)
    path = '/content/drive/MyDrive/NLP 2024/oh_algoritmo.txt'
except:
    print('Ignorando conexión drive-colab')



Mounted at /content/drive


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook


In [2]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("/content/drive/MyDrive/NLP 2024/oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrán: Jorge Drexler]
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?

[Estribillo: Jorge Drexler]
Dime qué debo cantar
Oh, algoritmo
Sé que lo sabes mejor
Incluso que yo mismo

[Verso 1: Nora Erez]
Wait, what's that money that you spent?
What's that sitting on your plate?
Do you want what you've been fed?
Are you the fish or bait?
Mmm, I'm on the top of the roof and I feel like a jail
Rather not pay the bail
To dangerous people with blood on their faces
So I'm sharing a cell with the masses
The underground always strive for the main
Streaming like Grande's big-ass ring
Screaming: I'll write you out my will
Conscious is free, but not the will
Conscious is free, but not the will
You

Fuente: https://genius.com/Jorge-drexler-oh-algoritmo-lyrics

### Pregunta 1.a (0.25 puntos)

Diseñe una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Es libre de elegir la forma de tokenizar mientras no utilice librerías con tokenizadores ya implementados. Puede utilizar la librería **re** importada para trabajar símbolos. Explique su razonamiento.


In [3]:
import re

In [4]:
def get_tokens2(texto):
  texto = re.sub('\W+',' ',texto)
  texto = re.sub('\d+','',texto)
  return texto.split()

In [5]:
def get_tokens(texto):
    ### Aquí inicia tu código ###
    tokens = re.split(r'\W+', texto)
    return [token for token in tokens if token != '']
    ### Aquí termina tu código ##

In [6]:
tokens = get_tokens(texto)
tokens

['Letra',
 'de',
 'Oh',
 'Algoritmo',
 'ft',
 'Nora',
 'Erez',
 'Refrán',
 'Jorge',
 'Drexler',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Estribillo',
 'Jorge',
 'Drexler',
 'Dime',
 'qué',
 'debo',
 'cantar',
 'Oh',
 'algoritmo',
 'Sé',
 'que',
 'lo',
 'sabes',
 'mejor',
 'Incluso',
 'que',
 'yo',
 'mismo',
 'Verso',
 '1',
 'Nora',
 'Erez',
 'Wait',
 'what',
 's',
 'that',
 'money',
 'that',
 'you',
 'spent',
 'What',
 's',
 'that',
 'sitting',
 'on',
 'your',
 'plate',
 'Do',
 'you',
 'want',


### Pregunta 1.b (0.25 puntos)
Explique su implementación aquí:
> La implementación de la función mediante el método **sub()** de la librería de expresiones regulares (**re**), "limpia" el texto o corpus (documento -cadena de palabras) de aquellos elementos gráficos presentes (paréntesis por ejemplo), de elementos contextuales del lenguaje que son propios del habla (inflexiones relacionadas a signos de exclamación, interrogación) y de redacción del texto (puntuación). Mediante el método **split()** el texto ya limpio de los elementos indicados, es llevado a minúscula y luego separado en las cadenas de texto que lo componen (palabras). La función retorna un documento tokenizado.

Implementación con la libreria NLTK

In [7]:
from nltk.tokenize import wordpunct_tokenize
nltk_tokens = wordpunct_tokenize(texto)
print(nltk_tokens)

['[', 'Letra', 'de', '"¡', 'Oh', ',', 'Algoritmo', '!"', 'ft', '.', 'Nora', 'Erez', ']', '[', 'Refrán', ':', 'Jorge', 'Drexler', ']', '¿', 'Quién', 'quiere', 'que', 'yo', 'quiera', 'lo', 'que', 'creo', 'que', 'quiero', '?', '¿', 'Quién', 'quiere', 'que', 'yo', 'quiera', 'lo', 'que', 'creo', 'que', 'quiero', '?', '¿', 'Quién', 'quiere', 'que', 'yo', 'quiera', 'lo', 'que', 'creo', 'que', 'quiero', '?', '¿', 'Quién', 'quiere', 'que', 'yo', 'quiera', 'lo', 'que', 'creo', 'que', 'quiero', '?', '¿', 'Quién', 'quiere', 'que', 'yo', 'quiera', 'lo', 'que', 'creo', 'que', 'quiero', '?', '¿', 'Quién', 'quiere', 'que', 'yo', 'quiera', 'lo', 'que', 'creo', 'que', 'quiero', '?', '[', 'Estribillo', ':', 'Jorge', 'Drexler', ']', 'Dime', 'qué', 'debo', 'cantar', 'Oh', ',', 'algoritmo', 'Sé', 'que', 'lo', 'sabes', 'mejor', 'Incluso', 'que', 'yo', 'mismo', '[', 'Verso', '1', ':', 'Nora', 'Erez', ']', 'Wait', ',', 'what', "'", 's', 'that', 'money', 'that', 'you', 'spent', '?', 'What', "'", 's', 'that', 's

### Pregunta 1.c (0.5 puntos)
¿Qué diferencias y similitudes encontrase al comparar la función de tokenización creada manualmente por ti contra la implementación de NLTK, al tokenizar la letra de la canción "Oh, algoritmo"?

**R.** La función de la librería NLTK si bien tokeniza el texto entregado, no elimina directamente expresiones de contexto o significado del texto (expresiones de exclamación, interrogación y otras como paréntesis o puntuación), u otros elementos presentes en el texto (como por ejemplo, números).

## P2. Stemming y Stopwords

En esta sección debera implementar funciones de stemming y stopwords basado en lo visto en clase. En la siguiente celda tiene el corpus que usara en esta sección:

In [8]:
# Corpus en español
corpus_espanol = [
    "¿Quién quiere que yo quiera lo que creo que quiero?",
    "Dime qué debo cantar",
    "Sé que lo sabes mejor"
]

# Corpus en inglés
corpus_ingles = [
    "What's that sitting on your plate?",
    "Do you want what you've been fed?",
    "Are you the fish or bait?"
]

### Pregunta 2.a (0.5 puntos)
Implemente una función **`get_vocab()`** que extraiga los tokens de un corpus. Puede utilizar la función de la sección anterior.

In [9]:
def get_vocab(corpus):
  corpus = " ".join(corpus)
  return list(set(get_tokens(corpus)))

In [10]:
vocab_esp = get_vocab(corpus_espanol)
print(vocab_esp)

['cantar', 'Quién', 'yo', 'lo', 'debo', 'mejor', 'Sé', 'quiera', 'creo', 'quiere', 'quiero', 'que', 'qué', 'Dime', 'sabes']


In [11]:
vocab_espanol = get_vocab(corpus_espanol)
len(vocab_espanol)

15

Resultado esperado (el orden puede variar):
```
['yo', 'debo', 'creo', 'Dime', 'lo', 'cantar', 'mejor', 'Sé', 'que', 'quiere', 'quiero', 'sabes', 'Quién', 'quiera', 'qué']
```

In [12]:
vocab_ing = get_vocab(corpus_ingles)
print(vocab_ing)

['plate', 'fed', 'Are', 'what', 's', 'want', 'bait', 'that', 've', 'fish', 'been', 'Do', 'on', 'you', 'your', 'What', 'or', 'sitting', 'the']


In [13]:
len(vocab_ing)

19

vocab_ingles = get_vocab(corpus_ingles)
vocab_ingles

Resultado esperado:
```
['fed', 'been', 'or', 'want', 'plate', 'the', 've', 'your', 's', 'you', 'what', 'Are', 'bait', 'What', 'fish', 'that', 'sitting', 'Do', 'on']
```

### Pregunta 2.b (0.5 puntos)
Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario preprocesado. Explique las reglas de stemming y elección de stopwords:

Explique sus reglas aquí:

Para las STOPWORDS se utilizó como primer criterio, la frecuencia de las palabras presentes en los corpus. De esta forma, las palabras de mayor frecuencias fueron eliminadas. Este primer criterio hace referencia a la ley de Zipf. Como segundo criterio se utilizó la extensión de las palabras. En castellano (y en inglés), muchas palabras de extensión reducida, además de ser de uso frecuente (lo que habla de poco contenido específico), corresponden a artículos, ilativos y proposiciones.

En el caso del STEMMING, se usaron reglas diferentes, según el idioma. Para el CASTELLANO, se suprimieron las vocales en caso que las palabras terminasen en vocal (como una forma de neutralizar el género); la letra 'S' en posición final (como una forma de neutralizar la referencia a la cantidad), y la terminación 'AR' en el caso de los verbos terminados en dichas letras.

En el caso del INGLÉS, el STEMMING definido suprimió las vocales y la letra 'R' en posiciones finales; la terminación 'ER' que hace referencia a los sujetos agentes (ej: singer, player, runner, walker, etc.), y la terminación 'ING' que es la terminación en ese idioma del gerundio verbal (singing, player, running, etc.)

In [14]:
def pre_processing2(documentos, vocabulario, idioma):
    """
    Función que realiza un preprocesamiento del lenguaje natural, que aplica reglas para eliminar stopwors y realizar stemmig sobre las palabras
    que integran los diferentes documentos del corpus.

    Argumentos
    -----------
    La función recibe como argumento un corpus (documentos), un vocabulario y un idioma (español o inglés)

    Resultado
    ----------
    El resultado de la función es un corpus tokenizado, donde los tokens o términos procesados, se almacenan en una lista

    """
    corpus = " ".join(documentos)
    corpus = get_tokens(corpus)
    lista = dict(zip(corpus,[corpus.count(palabra) for palabra in vocabulario])) # Conteo de las palabras del corpus con presencia en el diccionario
    lista = dict(sorted(lista.items(), key=lambda item:item[1],reverse=True)) # Jerarquización de las palabras del diccionario según su frecuencia en el corpus
    lista = [f'{k}' for k,v in lista.items() if lista[k] < 2] # Elimina palabras con frecuencias mayores.
    lista = [i for i in lista if len(i) > 3] # Se filtra palabras de menos de tres letras

    if idioma == 'espanol':
      nueva_lista = []
      for palabra in lista:
        if re.match(r'\w*[aeiouáéíóú]$', palabra): # Resta la vocal si la palabra termina en vocal (neutralizar el género de las palabras)
          palabra = palabra[:-1]
        elif palabra.endswith('s'): # Elimina la última letra si la palabra termina en 's' (plural de las palabras)
          palabra = palabra[:-1]
        elif palabra.endswith('ar'): # Elimina la terminación 'ar' de la palabra (indicativo de los verbos)
          palabra = palabra[:-2]

        nueva_lista.append(palabra)

    elif idioma == 'ingles':
      nueva_lista = []
      for palabra in lista:
        if re.match(r'\w*[aeiour]$', palabra): # Se elimina la vocal si la palabra termina en vocal y en 'r'.
          palabra = palabra[:-1]
        elif palabra.endswith('er'): # Se elimina la terminación 'er' (contrae el agente)
          palabra = palabra[:-2]
        elif palabra.endswith('ing'): # Se elimina terminación 'ing' (gerundio de los verbos)
          palabra = palabra[:-3]

        nueva_lista.append(palabra)

    return nueva_lista

In [15]:
def pre_processing(vocabulario, idioma):
    """
    Función que realiza un preprocesamiento del lenguaje natural, que aplica reglas para eliminar stopwors y realizar stemmig sobre las palabras
    que integran un vocabulario.

    Argumentos
    -----------
    La función recibe como argumento un vocabulario y un idioma (español o inglés)

    Resultado
    ----------
    El resultado de la función es un vocabulario tokenizado, donde los tokens o términos procesados, se almacenan en una lista

    """

    vocabulario = [i for i in vocabulario if len(i) > 3]

    if idioma == 'espanol':
      nueva_lista = []
      for palabra in vocabulario:
        if re.match(r'\w*[aeiouáéíóú]$', palabra): # Resta la vocal si la palabra termina en vocal (neutralizar el género de las palabras)
          palabra = palabra[:-1]
        elif palabra.endswith('s'): # Elimina la última letra si la palabra termina en 's' (plural de las palabras)
          palabra = palabra[:-1]
        elif palabra.endswith('ar'): # Elimina la terminación 'ar' de la palabra (indicativo de los verbos)
          palabra = palabra[:-2]

        nueva_lista.append(palabra)

    elif idioma == 'ingles':
      nueva_lista = []
      for palabra in vocabulario:
        if re.match(r'\w*[aeiour]$', palabra): # Se elimina la vocal si la palabra termina en vocal y en 'r'.
          palabra = palabra[:-1]
        elif palabra.endswith('er'): # Se elimina la terminación 'er' (contrae el agente)
          palabra = palabra[:-2]
        elif palabra.endswith('ing'): # Se elimina terminación 'ing' (gerundio de los verbos)
          palabra = palabra[:-3]

        nueva_lista.append(palabra)

    return nueva_lista

In [16]:
pre_processing2(corpus_ingles,vocab_ing,'ingles')

['What', 'that', 'sitt', 'you', 'plat', 'want', 'what', 'fish']

In [17]:
import nltk
from nltk.stem.porter import PorterStemmer
from nltk.stem.snowball import SnowballStemmer
from nltk.tokenize import word_tokenize
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [18]:
def pre_processing_nltk(corpus,idioma):
    """
    Función que realiza un preprocesamiento del lenguaje natural, que aplica las stopwors y el stemmig de la librería NLTK.

    Argumentos
    -----------
    La función recibe como argumento un corpus y un idioma (español o inglés)


    Resultado
    ----------
    El resultado de la función es un vocabulario tokenizado, donde los tokens o términos procesados, se almacenan en una lista
    """

    texto = " ".join(corpus)
    texto = re.sub('\W+',' ',texto)
    texto = re.sub('\d+','',texto)

    if idioma == 'english':
      porter = PorterStemmer()
      texto = [porter.stem(palabra) for palabra in texto.split()]
      stop = stopwords.words('english')

    elif idioma == 'spanish':
      stemmer = SnowballStemmer('spanish')
      texto = [stemmer.stem(palabra) for palabra in texto.split()]
      stop = stopwords.words('spanish')

    return [palabra for palabra in texto if palabra not in stop]

In [19]:
def pre_processing_nltk_2(vocabulario,idioma):
    """
    Función que realiza un preprocesamiento del lenguaje natural, que aplica las stopwors y el stemmig de la librería NLTK.

    Argumentos
    -----------
    La función recibe como argumento un vocabulario y un idioma (español o inglés)


    Resultado
    ----------
    El resultado de la función es un vocabulario tokenizado, donde los tokens o términos procesados, se almacenan en una lista
    """

    if idioma == 'english':
      porter = PorterStemmer()
      texto = [porter.stem(palabra) for palabra in vocabulario]
      stop = stopwords.words('english')

    elif idioma == 'spanish':
      stemmer = SnowballStemmer('spanish')
      texto = [stemmer.stem(palabra) for palabra in vocabulario]
      stop = stopwords.words('spanish')

    return [palabra for palabra in texto if palabra not in stop]

In [20]:
# Aplicar preprocesamiento a los vocabularios de ejemplo con NLTK

print('APLICACIÓN DE FUNCIONES PARA PRE-PROCESAMIENTO (STEMMING Y STOPWORS)\n')

print('1. Resultados stemming y stopwords con función propia (argumentos = vocabulario, idioma):')
vocab_procesado_espanol = pre_processing(vocab_esp,'espanol')
vocab_procesado_ingles = pre_processing(vocab_ing,'ingles')

print("a) Vocabulario función pre-procesado en español:", vocab_procesado_espanol)
print("b) Vocabulario función pre-procesado en inglés:", vocab_procesado_ingles, "\n")


print('2. Resultados stemming y stopwords con función basada en NLTK:')
vocab_procesado_espanol_nltk = pre_processing_nltk_2(vocab_esp, 'spanish')
vocab_procesado_ingles_nltk = pre_processing_nltk_2(vocab_ing, 'english')

# Mostrar resultados
print("c) Vocabulario procesado en español con NLTK:", vocab_procesado_espanol)
print("d) Vocabulario procesado en inglés con NLTK:", vocab_procesado_ingles, "\n")


APLICACIÓN DE FUNCIONES PARA PRE-PROCESAMIENTO (STEMMING Y STOPWORS)

1. Resultados stemming y stopwords con función propia (argumentos = vocabulario, idioma):
a) Vocabulario función pre-procesado en español: ['cant', 'Quién', 'deb', 'mejor', 'quier', 'cre', 'quier', 'quier', 'Dim', 'sabe']
b) Vocabulario función pre-procesado en inglés: ['plat', 'what', 'want', 'bait', 'that', 'fish', 'been', 'you', 'What', 'sitt'] 

2. Resultados stemming y stopwords con función basada en NLTK:
c) Vocabulario procesado en español con NLTK: ['cant', 'Quién', 'deb', 'mejor', 'quier', 'cre', 'quier', 'quier', 'Dim', 'sabe']
d) Vocabulario procesado en inglés con NLTK: ['plat', 'what', 'want', 'bait', 'that', 'fish', 'been', 'you', 'What', 'sitt'] 



## P3. Bag of Words (0.5 puntos)
Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:


In [21]:
d0 = 'El pájaro come semillas'
d1 = 'El pájaro se despierta y canta'
d2 = 'El pájaro canta y come semillas'
d3 = 'El pez come y nada en el agua'
d4 = 'El pez empieza a nadar'
d5 = 'El pez come alimento'
corpus = [d0, d1, d2, d3, d4, d5]

El objetivo da las siguientes secciones es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica **TF-IDF**.

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es la de **Bag of Words**, método mediante el cual se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que recibe como input un arreglo de documentos y devuelve un dataframe de pandas con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representan las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el BoW de un documento.

***Disclaimer: el orden de los resultados pueden variar.***


Por ejemplo para el siguiente corpus:

```
corpus = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |




In [22]:
import pandas as pd

Implementar función `bag_of_words()`

In [23]:
def bag_of_words(corpus):
    ### Aquí inicia tu código ###
    tokens = get_vocab(corpus)
    #tokens = pre_processing(tokens, 'espanol')
    df_corpus = pd.DataFrame([], columns=tokens)
    #Se recorre el corpus de manera secuencial, creando una lista de las apariciones en los tokens y agregandolas al pandas final
    for i in corpus:
        #Calculamos el largo del dataframe, para poder rellenarlo inicilamente con 0´s
        df_corpus.loc[len(df_corpus)] = pd.Series([0] * len(df_corpus.columns), index=df_corpus.columns)
        tokenizados = get_tokens(i)
        for k in tokenizados:
            if k in tokens:
                #index = df_corpus.index(-1)
                df_corpus.iloc[-1][k] += 1
    return df_corpus
    ### Aquí termina tu código ##

In [24]:
dataset_bow = bag_of_words(corpus)
dataset_bow

y  nada  agua  despierta  pez  se  semillas  canta  El  empieza  pájaro  \
0  0     0     0          0    0   0         1      0   1        0       1   
1  1     0     0          1    0   1         0      1   1        0       1   
2  1     0     0          0    0   0         1      1   1        0       1   
3  1     1     1          0    1   0         0      0   1        0       0   
4  0     0     0          0    1   0         0      0   1        1       0   
5  0     0     0          0    1   0         0      0   1        0       0   

   nadar  alimento  a  el  en  come  
0      0         0  0   0   0     1  
1      0         0  0   0   0     0  
2      0         0  0   0   0     1  
3      0         0  0   1   1     1  
4      1         0  1   0   0     0  
5      0         1  0   0   0     1

In [25]:
dataset_bow.shape

(6, 17)

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

## P4. TF-IDF

### 4.a TF (0.25 puntos)

Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia $\max_i({\text{tf}_{i,j}})$, donde
$i$ corresponde al índice de las filas (BoW) y $j$ al de las columnas (palabras). Es decir, dividir cada BoW sobre la cantidad de veces de la palabra que aparezca más veces en ese vector.


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{\max_i({\text{tf}_{i,j})}}$$

Implemente la función `calc_tf(dataset_bow)`, que entrega la matriz de TF normalizada del BoW del dataset.

In [26]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    for i in range(0, len(dataset_bow)):
        max_value = dataset_bow.iloc[i].max()
        dataset_bow.iloc[i] = dataset_bow.iloc[i]/max_value
    return dataset_bow
    ### Aquí termina tu código ##

In [27]:
tf = calc_tf(dataset_bow)
tf

y  nada  agua  despierta  pez  se  semillas  canta  El  empieza  pájaro  \
0  0     0     0          0    0   0         1      0   1        0       1   
1  1     0     0          1    0   1         0      1   1        0       1   
2  1     0     0          0    0   0         1      1   1        0       1   
3  1     1     1          0    1   0         0      0   1        0       0   
4  0     0     0          0    1   0         0      0   1        1       0   
5  0     0     0          0    1   0         0      0   1        0       0   

   nadar  alimento  a  el  en  come  
0      0         0  0   0   0     1  
1      0         0  0   0   0     0  
2      0         0  0   0   0     1  
3      0         0  0   1   1     1  
4      1         0  1   0   0     0  
5      0         1  0   0   0     1

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

### 4.b IDF (0.5 puntos)

Implementar `calc_idf(dataset_bow)`. Ésta debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el cálculo de cada idf por palabra.

Recordar que $\text{idf}_{t_i} = \log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ número de documentos que contienen la palabra $t_i$

In [28]:
import numpy as np

In [29]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    N=len(dataset_bow)
    lista_columnas = dataset_bow.columns
    dicc_idf = {}
    for i in lista_columnas:
        n_i = dataset_bow[i].sum()
        idf_i = np.log(N/n_i) / np.log(10)
        dicc_idf[i]= idf_i
    return dicc_idf
    ### Aquí termina tu código ##

In [30]:
idf = calc_idf(dataset_bow)
idf

{'y': 0.30102999566398114,
 'nada': 0.7781512503836435,
 'agua': 0.7781512503836435,
 'despierta': 0.7781512503836435,
 'pez': 0.30102999566398114,
 'se': 0.7781512503836435,
 'semillas': 0.47712125471966244,
 'canta': 0.47712125471966244,
 'El': 0.0,
 'empieza': 0.7781512503836435,
 'pájaro': 0.30102999566398114,
 'nadar': 0.7781512503836435,
 'alimento': 0.7781512503836435,
 'a': 0.7781512503836435,
 'el': 0.7781512503836435,
 'en': 0.7781512503836435,
 'come': 0.17609125905568124}

Solución esperada:
```
{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'come': 0.17609125905568124,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'se': 0.7781512503836436,
 'en': 0.7781512503836436,
 'y': 0.3010299956639812,
 'alimento': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'empieza': 0.7781512503836436,
 'canta': 0.47712125471966244,
 'agua': 0.7781512503836436,
 'nada': 0.7781512503836436}
 ```

### 4.c TF-IDF (0.25 puntos)
Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [31]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    for i in tf.columns:
        tf[i] = tf[i] * idf[i]
    return tf
    ### Aquí termina tu código ##

In [32]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

y      nada      agua  despierta      pez        se  semillas  \
0  0.00000  0.000000  0.000000   0.000000  0.00000  0.000000  0.477121   
1  0.30103  0.000000  0.000000   0.778151  0.00000  0.778151  0.000000   
2  0.30103  0.000000  0.000000   0.000000  0.00000  0.000000  0.477121   
3  0.30103  0.778151  0.778151   0.000000  0.30103  0.000000  0.000000   
4  0.00000  0.000000  0.000000   0.000000  0.30103  0.000000  0.000000   
5  0.00000  0.000000  0.000000   0.000000  0.30103  0.000000  0.000000   

      canta   El   empieza   pájaro     nadar  alimento         a        el  \
0  0.000000  0.0  0.000000  0.30103  0.000000  0.000000  0.000000  0.000000   
1  0.477121  0.0  0.000000  0.30103  0.000000  0.000000  0.000000  0.000000   
2  0.477121  0.0  0.000000  0.30103  0.000000  0.000000  0.000000  0.000000   
3  0.000000  0.0  0.000000  0.00000  0.000000  0.000000  0.000000  0.778151   
4  0.000000  0.0  0.778151  0.00000  0.778151  0.000000  0.778151  0.000000   
5  0.000000  0.0  0.000000  0.00000  0.000000  0.778151  0.000000  0.000000   

         en      come  
0  0.000000  0.176091  
1  0.000000  0.000000  
2  0.000000  0.176091  
3  0.778151  0.176091  
4  0.000000  0.000000  
5  0.000000  0.176091

Solución esperada:

|    |   El |   pájaro |   despierta |       el |     come |        a |    nadar |       se |       en |       y |   alimento |   semillas |     pez |   empieza |    canta |     agua |     nada |
|:---|-----:|---------:|------------:|---------:|---------:|---------:|---------:|---------:|---------:|--------:|-----------:|-----------:|--------:|----------:|---------:|---------:|---------:|
| d0 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0        |   0.477121 | 0       |  0        | 0        | 0        | 0        |
| d1 |    0 |  0.30103 |    0.778151 | 0        | 0        | 0        | 0        | 0.778151 | 0        | 0.30103 |   0        |   0        | 0       |  0        | 0.477121 | 0        | 0        |
| d2 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0.30103 |   0        |   0.477121 | 0       |  0        | 0.477121 | 0        | 0        |
| d3 |    0 |  0       |    0        | 0.778151 | 0.176091 | 0        | 0        | 0        | 0.778151 | 0.30103 |   0        |   0        | 0.30103 |  0        | 0        | 0.778151 | 0.778151 |
| d4 |    0 |  0       |    0        | 0        | 0        | 0.778151 | 0.778151 | 0        | 0        | 0       |   0        |   0        | 0.30103 |  0.778151 | 0        | 0        | 0        |
| d5 |    0 |  0       |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0.778151 |   0        | 0.30103 |  0        | 0        | 0        | 0        |

## P5. Cosine-similarity (0.25 puntos)
Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica.

Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos. Concluya cuáles son los dos documentos más similares.

In [33]:
def cosine_similarity(v1, v2):
    ### Aquí inicia tu código ###
    n1 = np.dot(v1, v2)
    n2 = np.linalg.norm(v1)* np.linalg.norm(v2)
    return n1/n2
    ### Aquí termina tu código ##

In [34]:
similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity

for i in range(6):
  mask = [k != i for k in range(6)]
  j = np.argmax(similarity_matrix[i][mask])

  print(corpus[i])
  print("> Mas similar:", np.array(corpus)[mask][j])
  print("> Similitud:", similarity_matrix[i][mask][j], "\n")

El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414 

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945 

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414 

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406166 

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.0769507840675271 

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.08787900372173231 



Solución esperada:
```
El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323
```

## P6 N-gramas (0.75 punto)

En esta sección debera determinar los n-gramas del la cancion "Oh algoritmo".

### 6.a Corpus de entrenamiento y test (0.25 puntos)

En esta subsección debera definir el conjunto de entrenamiento y test de un corpus. Eliga una particion del 80% y 20% del texto.

In [35]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

El archivo no se encuentra.


Defina una funcion `get_sentences()` que entregue todas las oraciones del corpus que contengan al menos una palabra.

In [36]:
def get_sentences(texto):
  ## Implementar aquí
  lines = texto.splitlines()
  sentences_list = []

  for line in lines:
    if line != '':
      sentences_list.append(line)
  return sentences_list
  ##

In [37]:
oraciones_limpias = get_sentences(texto)
oraciones_limpias

['[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]',
 '[Refrán: Jorge Drexler]',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '[Estribillo: Jorge Drexler]',
 'Dime qué debo cantar',
 'Oh, algoritmo',
 'Sé que lo sabes mejor',
 'Incluso que yo mismo',
 '[Verso 1: Nora Erez]',
 "Wait, what's that money that you spent?",
 "What's that sitting on your plate?",
 "Do you want what you've been fed?",
 'Are you the fish or bait?',
 "Mmm, I'm on the top of the roof and I feel like a jail",
 'Rather not pay the bail',
 'To dangerous people with blood on their faces',
 "So I'm sharing a cell with the masses",
 'The underground always strive for the main',
 "Streaming like Grande's big-ass ring",
 "Screaming: I'll wr

Debería obtener en total 87 oraciones.

In [38]:
len(oraciones_limpias) == 87

True

Ahora definiremos el conjunto de entrenamiento y prueba para las oraciones:

In [39]:
split = int(len(oraciones_limpias) * 0.8)
train_corpus = oraciones_limpias[:split]
test_corpus = oraciones_limpias[split:]

len(train_corpus), len(test_corpus)

(69, 18)

### 6.b Estimación de N-gramas (0.5 puntos)

Defina una función que reciba una lista de oraciones de un corpus y un N que indique el tamaño de los N-gramas. La función debe retornar un diccionario de Python donde la llave es un token (o palabra) y el valor es la cantidad de veces que ocurre el token, es decir, la frecuencia. En el caso de N-gramas con N mayor a 1 (como bi-gramas o tri-gramas) debe añadir un token especial al inicio o final de cada oración según corresponda (ver clases del curso).

In [40]:
from nltk.tokenize import word_tokenize

In [41]:
def n_grams(corpus, n=3):
  ## Implementar aquí
  # Diccionario para almacenar la frecuencia de los tokens
  frequency_dict = {}

  # Iterar sobre cada oración en el corpus
  for i in corpus:
    #Revisar acá, cambiar por word_tokenize
    tokens = get_tokens(i)

    # Agregar tokens especiales al inicio y final de la oración para n-gramas con n > 1
    tokens = ['<s>'] * n + tokens + ['</s>']


    # Generar n-gramas
    for i in range(len(tokens) - n + 1):
      ngram = tuple(tokens[i:i + n])

      # Actualizar la frecuencia del n-grama en el diccionario
      if ngram in frequency_dict:
        frequency_dict[ngram] += 1
      else:
        frequency_dict[ngram] = 1

  return frequency_dict

In [42]:
n_grams(train_corpus, n=1)

{('<s>',): 69,
 ('Letra',): 1,
 ('de',): 3,
 ('Oh',): 3,
 ('Algoritmo',): 1,
 ('ft',): 1,
 ('Nora',): 3,
 ('Erez',): 3,
 ('</s>',): 69,
 ('Refrán',): 2,
 ('Jorge',): 10,
 ('Drexler',): 10,
 ('Quién',): 11,
 ('quiere',): 11,
 ('que',): 38,
 ('yo',): 14,
 ('quiera',): 11,
 ('lo',): 13,
 ('creo',): 11,
 ('quiero',): 11,
 ('Estribillo',): 3,
 ('Dime',): 3,
 ('qué',): 3,
 ('debo',): 3,
 ('cantar',): 3,
 ('algoritmo',): 3,
 ('Sé',): 2,
 ('sabes',): 2,
 ('mejor',): 2,
 ('Incluso',): 2,
 ('mismo',): 2,
 ('Verso',): 4,
 ('1',): 1,
 ('Wait',): 1,
 ('what',): 3,
 ('s',): 3,
 ('that',): 4,
 ('money',): 1,
 ('you',): 6,
 ('spent',): 1,
 ('What',): 1,
 ('sitting',): 1,
 ('on',): 3,
 ('your',): 1,
 ('plate',): 1,
 ('Do',): 1,
 ('want',): 4,
 ('ve',): 1,
 ('been',): 1,
 ('fed',): 1,
 ('Are',): 1,
 ('the',): 8,
 ('fish',): 1,
 ('or',): 1,
 ('bait',): 1,
 ('Mmm',): 1,
 ('I',): 7,
 ('m',): 2,
 ('top',): 1,
 ('of',): 1,
 ('roof',): 1,
 ('and',): 1,
 ('feel',): 1,
 ('like',): 3,
 ('a',): 3,
 ('jail',): 1,


In [43]:
n_grams(train_corpus, n=2)

{('<s>', '<s>'): 69,
 ('<s>', 'Letra'): 1,
 ('Letra', 'de'): 1,
 ('de', 'Oh'): 1,
 ('Oh', 'Algoritmo'): 1,
 ('Algoritmo', 'ft'): 1,
 ('ft', 'Nora'): 1,
 ('Nora', 'Erez'): 3,
 ('Erez', '</s>'): 3,
 ('<s>', 'Refrán'): 2,
 ('Refrán', 'Jorge'): 2,
 ('Jorge', 'Drexler'): 10,
 ('Drexler', '</s>'): 10,
 ('<s>', 'Quién'): 11,
 ('Quién', 'quiere'): 11,
 ('quiere', 'que'): 11,
 ('que', 'yo'): 13,
 ('yo', 'quiera'): 11,
 ('quiera', 'lo'): 11,
 ('lo', 'que'): 11,
 ('que', 'creo'): 11,
 ('creo', 'que'): 11,
 ('que', 'quiero'): 11,
 ('quiero', '</s>'): 11,
 ('<s>', 'Estribillo'): 2,
 ('Estribillo', 'Jorge'): 2,
 ('<s>', 'Dime'): 3,
 ('Dime', 'qué'): 3,
 ('qué', 'debo'): 3,
 ('debo', 'cantar'): 3,
 ('cantar', '</s>'): 3,
 ('<s>', 'Oh'): 2,
 ('Oh', 'algoritmo'): 2,
 ('algoritmo', '</s>'): 2,
 ('<s>', 'Sé'): 2,
 ('Sé', 'que'): 2,
 ('que', 'lo'): 2,
 ('lo', 'sabes'): 2,
 ('sabes', 'mejor'): 2,
 ('mejor', '</s>'): 2,
 ('<s>', 'Incluso'): 2,
 ('Incluso', 'que'): 2,
 ('yo', 'mismo'): 2,
 ('mismo', '</s>'):

In [44]:
n_grams(train_corpus, n=3)

{('<s>', '<s>', '<s>'): 69,
 ('<s>', '<s>', 'Letra'): 1,
 ('<s>', 'Letra', 'de'): 1,
 ('Letra', 'de', 'Oh'): 1,
 ('de', 'Oh', 'Algoritmo'): 1,
 ('Oh', 'Algoritmo', 'ft'): 1,
 ('Algoritmo', 'ft', 'Nora'): 1,
 ('ft', 'Nora', 'Erez'): 1,
 ('Nora', 'Erez', '</s>'): 3,
 ('<s>', '<s>', 'Refrán'): 2,
 ('<s>', 'Refrán', 'Jorge'): 2,
 ('Refrán', 'Jorge', 'Drexler'): 2,
 ('Jorge', 'Drexler', '</s>'): 10,
 ('<s>', '<s>', 'Quién'): 11,
 ('<s>', 'Quién', 'quiere'): 11,
 ('Quién', 'quiere', 'que'): 11,
 ('quiere', 'que', 'yo'): 11,
 ('que', 'yo', 'quiera'): 11,
 ('yo', 'quiera', 'lo'): 11,
 ('quiera', 'lo', 'que'): 11,
 ('lo', 'que', 'creo'): 11,
 ('que', 'creo', 'que'): 11,
 ('creo', 'que', 'quiero'): 11,
 ('que', 'quiero', '</s>'): 11,
 ('<s>', '<s>', 'Estribillo'): 2,
 ('<s>', 'Estribillo', 'Jorge'): 2,
 ('Estribillo', 'Jorge', 'Drexler'): 2,
 ('<s>', '<s>', 'Dime'): 3,
 ('<s>', 'Dime', 'qué'): 3,
 ('Dime', 'qué', 'debo'): 3,
 ('qué', 'debo', 'cantar'): 3,
 ('debo', 'cantar', '</s>'): 3,
 ('<s>',

Debe mostrar que su método funciona para $N = 1,2,3$.

## P7. Perplexity (1 punto)

En esta sección evaluarán su modelo de n-gramas y determinarán la probabilidad de oraciones y la perplejidad con un conjunto de test. Recuerde que la perplejidad se define de la siguiente manera:

$$
\text{Perplexity} = 2^{-l} \quad \quad l = \frac{1}{M} \sum_{i=1}^{m} \log p(s_i)
$$

con $m$ el número de oraciones del corpus y $M$ el tamaño del vocabulario.

### 7.a Obtener probabilidades (0.5 puntos)

En esta sección implementará una función que determine la probabilidad de una oración.

Defina una función que reciba una oración, un diccionario con n-gramas y el valor de $n$. La función debe entregar la probabilidad de cualquier oración.

**Hint**: No olvide los posibles casos borde, como palabras fuera del vocabulario.

In [45]:
def get_probability(sentence, n_grams_frequency, n):
  ## Implementar aquí
  probability = 1
  tokens = get_tokens(sentence)
  n_tokens = len(tokens)
  # Se añade padding y STOP
  tokens = ['<s>'] * (n - 1) + tokens + ['</s>']

  # Se recorren los n-gramas
  for i in range(n_tokens + 1):
    n_gram = tuple(tokens[i:i+n])
    n_1_gram = tuple(tokens[i:i+n-1])

    # Frecuencias para n-grama y (n-1)-grama
    n_gram_freq = n_grams_frequency[n_gram] if n_grams_frequency.get(n_gram) else 0
    n_1_gram_freq = n_grams_frequency[n_1_gram] if n_grams_frequency.get(n_1_gram) else 0

    # Se ignora si el n-grama no se conoce
    if n_gram_freq == 0: continue

    probability *= n_gram_freq / n_1_gram_freq

  return probability

Pruebe su función con oraciones frecuentes y comente sus resultados



In [46]:
n = 3

n_gram = n_grams(train_corpus, n)
n_1_gram = n_grams(train_corpus, n-1)

n_gram.update(n_1_gram)

print(get_probability('Quién quiere que yo quiera lo que creo que quiero', n_gram, n))
print(get_probability('El perro ladra fuerte', n_gram, n))
print(get_probability('elegiste letra final cantar', n_gram, n))

0.13489409141583053
1
1


### 7.b Perplexity en conjunto de test (0.5 puntos)

En esta sub-sección deberá calcular la perplejidad del corpus de test.

Defina una función que reciba un corpus de test y retorne la perplexity (ver clases del curso). Utilice la función de la sección anterior.

In [47]:
def corpus_size(corpus):
  size = 0
  for sentence in corpus:
    tokens = get_tokens(sentence)
    size += len(tokens)
  return size

In [48]:
def get_perplexity(corpus, n):
  ## implementar aquí
  n_grams_frequency = n_grams(train_corpus, n)
  n_1_grams_frequency = n_grams(train_corpus, n-1) if n > 1 else {(): corpus_size(train_corpus)}
  n_grams_frequency.update(n_1_grams_frequency)

  M = corpus_size(corpus)
  sum_prob = 0
  for sentence in corpus:
    sum_prob += np.log2(get_probability(sentence, n_grams_frequency, n))

  l = sum_prob / M

  return 2**(-l)
  ##

In [49]:
get_perplexity(test_corpus, 3)

1.46110496358599

In [54]:
len(get_vocab(test_corpus))

20

Dé una interpretacion de la perplexity en el corpus de test:

**RESPUESTA:** En términos generales, la perplejidad de un modelo de lenguaje contrasta un modelo de lenguaje de probabilidad uniforme, donde la perplejidad de base es igual al tamaño del vocabulario del corpus, con un modelo de lenguaje que aprovecha las propiedades estadísticas del lenguaje, es decir, que tiene la capacidad de asignar mayores probabilidades a oraciones (n-gramas) que son más frecuentes dentro de un corpus. En esa dirección, la perplejidad de un modelo de distribución uniforme, para el corpus de test, es igual a 20 (que es la extensión del vocabulario que se puede extraer de él). Si dicha cifra la contrastamos con la perplejidad obtenida por el modelo de tri-grama entrenado (1.46), se puede decir que el modelo de lenguaje entrenado a partir de corpus de train, es bastante mejor en términos de que asigna probabilidades más altas de determinadas oraciones que a otras en el corpus de test. Sin embargo, es necesario decir que la baja perplejidad obtenida por el modelo entrenado, también puede explicarse en la medida que las oraciones presentes en el corpus de test, corresponden al coro de la canción, las que también se encuentran en el corpus de train.

## P8. Interpolación Lineal (0.5 puntos)

Cree una función que obtenga la probabilidad de una oración interpolando linealmente modelos de unigrama, bigrama y trigrama ponderados por $\lambda_1, \lambda_2$ y $\lambda_3$ respectivamente. Para esto use las funciones que creó anteriormente.

In [69]:
def get_probability_lineal_interpol(sentence, corpus, l_1, l_2, l_3):
  ## Implementar aquí
  n_grams_frequency = {(): corpus_size(corpus)}
  for n in range(1, 4):
    n_grams_frequency.update(n_grams(corpus, n))

  probability = 1
  tokens = get_tokens(sentence)
  n_tokens = len(tokens)
  # Se añade padding y STOP
  tokens = ['<s>', '<s>'] + tokens + ['</s>']

  for i in range(n_tokens + 1):

    # Se calculan los 3-grama, 2-grama y 1-grama
    n_gram_1 = tuple(tokens[i:i+3])
    n_1_gram_1 = tuple(tokens[i:i+2])

    n_gram_2 = tuple(tokens[i+1:i+3])
    n_1_gram_2 = tuple(tokens[i+1:i+2])

    n_gram_3 = tuple(tokens[i+2:i+3])
    n_1_gram_3 = tuple(tokens[i+3:i+3])

    # Se calculan las frecuencias de cada grama
    n_gram_1_freq = n_grams_frequency[n_gram_1] if n_grams_frequency.get(n_gram_1) else 0
    n_1_gram_1_freq = n_grams_frequency[n_1_gram_1] if n_grams_frequency.get(n_1_gram_1) else 0

    n_gram_2_freq = n_grams_frequency[n_gram_2] if n_grams_frequency.get(n_gram_2) else 0
    n_1_gram_2_freq = n_grams_frequency[n_1_gram_2] if n_grams_frequency.get(n_1_gram_2) else 0

    n_gram_3_freq = n_grams_frequency[n_gram_3] if n_grams_frequency.get(n_gram_3) else 0
    n_1_gram_3_freq = n_grams_frequency[n_1_gram_3] if n_grams_frequency.get(n_1_gram_3) else 0

    # Se calculan las probabilidades de cada grama
    n1_prob = n_gram_1_freq / n_1_gram_1_freq if n_gram_1_freq != 0 else 0
    n2_prob = n_gram_2_freq / n_1_gram_2_freq if n_gram_2_freq != 0 else 0
    n3_prob = n_gram_3_freq / n_1_gram_3_freq if n_gram_3_freq != 0 else 0

    # Si la probabilidad es 0, se pasa al siguiente
    if ((n1_prob * l_1) + (n2_prob * l_2) + (n3_prob * l_3)) == 0: continue

    probability *= (n1_prob * l_1) + (n2_prob * l_2) + (n3_prob * l_3)

  return probability
  ##

Defina una función para calcular la perplejidad de un corpus con interpolación lineal.

In [70]:
def get_pp_interpol(corpus, l_1, l_2, l_3):
  ## Implementar aquí
  M = corpus_size(corpus)
  sum_prob = 0
  for sentence in corpus:
    sum_prob += np.log2(get_probability_lineal_interpol(sentence, train_corpus, l_1, l_2, l_3))

  l = sum_prob / M

  return 2**(-l)
  ##

Ahora haga pruebas con distintos valores de $\lambda_1, \lambda_2$ y $\lambda_3$, incluyendo valores extremos (por ejemplo $[1, 0, 0]$). Comente sus resultados.

In [72]:
print("Perplexity Interpolación uni-grama:",get_pp_interpol(test_corpus, 0, 0, 1).round(4))
print("Perplexity Interpolación tri-grama:",get_pp_interpol(test_corpus, 1, 0, 0).round(4))
print("Perplexity Interpolación para bi-grama:",get_pp_interpol(test_corpus, 0, 1, 0).round(4))
print("Perplexity Interpolación equi-ponderada",get_pp_interpol(test_corpus, 0.33, 0.33, 0.34).round(4))

Perplexity Interpolación uni-grama: 38.9702
Perplexity Interpolación tri-grama: 1.4611
Perplexity Interpolación para bi-grama: 2.1914
Perplexity Interpolación equi-ponderada 2.6197


Análisis:

Los resultados de la interpolación revela por un lado, que el modelo queda bien definido, resolviendo de forma adecuada el problema del sparness que deben afrontar los modelos probabilísticos de lenguaje (que alguna oración no esté contenida en el corpus de entrenamiento y que sí aparezca en el corpus de test, lo que llevaría a que el modelo quedase indefinido; es decir, no podría asignar una probabilidad a una determinada oración que no haya visto en el corpus de test). Los resultados arrojados para la perplexity según hiperparámetros extremos (donde sólo a un lambda se asigna el valor de 1, y cero para los demás -dado que por definición, la suma de los hiperparámetros es 1), revela la perplexity más alta para el modelo de uni-grama, la más baja para el modelo de tri-gramas, y una perplexity un poco más alta para el modelo de bi-gramas. Al asignar valores uniformes a los hiperparámetros, se observa puede observar una perplexity un poco más alta que para el modelo de bi-gramas, pero aún baja, en comparación a la perplexity del modelo de uni-grama.